### ISSUE TITLE
Upgrade to TF 2.6

### ISSUE THREAD

**Issue by dakshvar22** [RasaHQ/rasa#7619]
TF `2.4` is out and hence we should update the dependency inside Rasa OS to use it.
Introduces some breaking changes as mentioned in the [changelog](https://github.com/tensorflow/tensorflow/releases)

Update: we are now targeting 2.6 see [here](https://rasa-hq.slack.com/archives/C8LEHMTAM/p1631797167445400?thread_ts=1631788033.438000&cid=C8LEHMTAM)

Note that there's a draft PR [here](https://github.com/RasaHQ/rasa/pull/8899) that already contains some of the necessary changes. Known remaining tasks in order to actually update code for TF `2.5` are (this list contains only the known remaining tasks, there may be others hence it is *not* an exhaustive of all remaining tasks):
- [x] Run model regression tests on all configs and datasets to see if training times / accuracy are the same
- [x] see TODOs marked with 2.5 (ex: [here](https://github.com/RasaHQ/rasa/pull/8899/files#diff-1b0d53e90a7f1dcca1113cda22c58c98685004f8dfc7ba3d4e5bf9cee78e6a19R253))
- [x] #9433
- [x] #9512

Temporarily de-prioritised followups:
- [ ] #9514 
- [ ] #9515 
- [ ] #9516 
- [ ] #9764
- [ ] https://github.com/RasaHQ/rasa/issues/9666
- [ ] https://github.com/RasaHQ/rasa/issues/9798
- [ ] https://github.com/RasaHQ/rasa/issues/9734

**Definition of done**
~No memory tests failing~
~No increased timeouts for tests~
~Model regression tests passing with approx. same time and accuracy~
- [X] Followup issue created: determine w DevRel how to communicate hits to community https://github.com/RasaHQ/research/issues/220
- [X] Followup issue created: determine w CSE how to communicate hits to customers https://github.com/RasaHQ/research/issues/219
- [X] Followup issue created: update CUDA drivers on GPU runners to be compatible with TF 2.5 : https://github.com/RasaHQ/rasa/issues/9611
- [X] Model regression tests on GPU have approx. same accuracy
- [X] Lockfile reflects correct TF dependency
- [X] Create a PR for the bumped up dependency
- [X] Create followup issue to investigate and address memory problems on Windows CI tests, so that [this fix](https://github.com/RasaHQ/rasa/pull/9733) can be reverted https://github.com/RasaHQ/rasa/issues/9734
- [X] Create followup issue to investigate and address memory problems for Hermit CI model regression tests so that they can be run on schedule again
https://github.com/RasaHQ/rasa/issues/9798

**Comment by eikooc** [comment]
Rasa does not work on the new M1 chip before upgrading this dependency. Just thought I would leave that as a note here.

**Comment by Ghostvv** [comment]
blocked until a new version is released with this fix: https://github.com/tensorflow/tensorflow/pull/45534/files#diff-39eadfc771b57ef11bff67b85cc38a47b9c13ec203b145e3feaf556a43bdccb1R4762

**Comment by Ghostvv** [comment]
pushed the necessary code updates to `tf-2.4` branch

**Comment by Ghostvv** [comment]
blocked by: https://github.com/tensorflow/tensorflow/issues/46511

**Comment by koernerfelicia** [comment]
Once this is done should check this: https://github.com/RasaHQ/rasa/issues/7762

**Comment by Ghostvv** [comment]
most probably we have to wait till 2.5

**Comment by koernerfelicia** [comment]
Related: https://github.com/RasaHQ/rasa/issues/7793

**Comment by luzhongqiu** [comment]
same issue, wait for tf above 2.4 version, thanks

**Comment by rmiaouh** [comment]
Hi guys, same issue, we will need tf  2.4 version, ty for all.

**Comment by koernerfelicia** [comment]
Once this is done we should revisit this and see if it has been fixed:

https://github.com/RasaHQ/rasa/issues/8004

**Comment by Ghostvv** [comment]
we cannot upgrade to `2.4`, we need to wait till `2.5`

**Comment by koernerfelicia** [comment]
Yeah, sorry I was being unclear, at this point when I say "done" I mean whatever tensorflow update we eventually manage

**Comment by mpomarole** [comment]
2.5 is now out!

https://developer.apple.com/metal/tensorflow-plugin/

https://github.com/tensorflow/tensorflow/releases/tag/v2.5.0

Seems like this can move forward?

**Comment by tmbo** [comment]
@TyDunn is this one still blocked? given that we'll need to buy m1s as well, would be great to sort this out 😅

**Comment by koernerfelicia** [comment]
@tmbo I'm trying to verify this atm, will update when I know more!

**Comment by mpomarole** [comment]
I am available to help migrate this over and test. Let me know if there is any ongoing work and how I can help speed it up. I have a M1 on hand and the desire to get rasa working asap on it :)

**Comment by koernerfelicia** [comment]
@mpomarole thank you for the kind offer! We'll let you know if we need any support :)

**Comment by alopez** [comment]
@aeshky primary assignee
@dakshvar22 reviewer

**Comment by vincelamm** [comment]
> I am available to help migrate this over and test. Let me know if there is any ongoing work and how I can help speed it up. I have a M1 on hand and the desire to get rasa working asap on it :)

Same here. Let me know, if you need support in testing.

**Comment by samsucik** [comment]
I'm now picking this one up 🙂  Will let you know if there's space for help (esp. for testing on M1s), @mpomarole @vincelamm.

**Comment by m-vdb** [comment]
@samsucik in case you need a review from / want to discuss with Enable, @twerkmeister is the person you can reach out to. This week is already a bit busy for him, but he'll make sure that early next week he has the time to help you. (if something urgent comes up, we can figure it out together). cc @TyDunn

**Comment by samsucik** [comment]
**About the peculiar `keras-nightly` dependency**

In short, [TF2.5 introduced this `keras-nightly` dependency](https://github.com/tensorflow/tensorflow/commit/d171d94d899cc73e18bd8b2b11b5c320ea7aac2c#diff-f526feeafa1000c4773410bdc5417c4022cb2c7b686ae658b629beb541ae9112R107) for nebulous reasons which may not even be important in our context. Today, installing Rasa doesn't install `keras` or `keras-nightly` at all, so this would be a peculiar addition to our dependencies! 🤨 

Besides bringing a "nightly" thing into everything that uses tensorflow (a bit scary!), this change also means that when you now install `tensorflow`, you get 2 options for Keras: `import keras` vs `from tensorflow import keras`. This is actually OK as long as you remember to **always use the tensorflow Keras** -- as [recommended by the Keras authors](https://github.com/keras-team/keras/releases/tag/2.4.0) and as [done everywhere inside the TF repo (checked also by me)](https://www.gitmemory.com/issue/tensorflow/tensorflow/49298/844058329).

Some background I've been able to collect: A standalone keras package [was once a thing](https://pypi.org/project/keras/2.4.3/) but then the tensorflow Keras [became the preferred (and supported) option](https://github.com/keras-team/keras/releases/tag/2.4.0). The [plan is to return to having standalone Keras](https://github.com/tensorflow/tensorflow/commit/d171d94d899cc73e18bd8b2b11b5c320ea7aac2c#diff-f526feeafa1000c4773410bdc5417c4022cb2c7b686ae658b629beb541ae9112R105), which is perhaps why `tensorflow` includes `keras-nightly` in preparation for including `keras` in the future. Meanwhile, the `keras` package [is kind of abandoned with the last version being 2.4.3](https://pypi.org/project/keras/2.4.3/), but release candidates 2.5, 2.6 and even 2.7 live happily both for [`keras`](https://pypi.org/project/keras/#history) and for [`keras-nightly`](https://pypi.org/project/keras-nightly/#history) and seem to be more or less keeping up with the `tensorflow` releases, preparing the ground for a return of a stable, up-to-date `keras` package.


**What do we do?**

I think it makes sense to overlook/ignore the `keras-nightly` dependency in the interest of using TF 2.5. Possible concerns:
- If someone installs Rasa and wants to use an older version of `keras` (e.g. 2.4.3): They'll have to override `keras-nightly` (which gets installed as part of Rasa's dependencies) with whatever other `keras` they want. It will not be in line with Tensorflow/Keras recommendations (or with ours), and it will be very much something caused by Tensorflow and by their decision to include `keras-nightly`, not by us.
- If someone's scared of having a nightly package in their installation: So are we! But the package should actually be safe to remove post-hoc (on the basis of not being used within Tensorflow's or our code at all), without any adverse effects. Keras should be used like `from tensorflow import keras` anyway, and should not be needed at all (as a standalone package).

**Comment by samsucik** [comment]
@m-vdb thanks, I won't need a review this week. This is far from being ready for one. Next week, though, some help would be great 🙂

**Comment by samsucik** [comment]
**On removing `@training.enable_multi_worker` ([this TODO item](https://github.com/RasaHQ/rasa/blob/044e47d671148f64634b1250d50d329c1731d4b9/rasa/utils/tensorflow/temp_keras_modules.py#L39))**

This decorator isn't used since TF 2.4 (removed [here](https://github.com/tensorflow/tensorflow/commit/ccdb409311c964dae895b04ffdd26bfa3ad60b0b)). It was in our code only because we copy-pasted the function signature of `Model.fit` which, at the time, included the decorator, but [this was later removed in `tensorflow`'s code](https://github.com/tensorflow/tensorflow/commit/ccdb409311c964dae895b04ffdd26bfa3ad60b0b#diff-f672b0b39d367d25ef24af024596744c7bd76cd77a49f6fe712c4569202cd2cdL827). In order to really enable distributed training, we need [take a proper look at the recommended ways of doing it](https://www.tensorflow.org/tutorials/distribute/multi_worker_with_keras) -- and that's out of the scope of this issue 🙂

**Comment by samsucik** [comment]
**Performance hits with TF 2.5**

I now started looking into some failing Windows tests (see e.g. the two failures in [this CI check run](https://github.com/RasaHQ/rasa/actions/runs/1029867150)), and I quickly discovered that training `e2ebot` takes much longer with TF 2.5 (the `tf-2.5` branch, which I keep up to date with `main`) compared to the current `main` branch. When I locally time the execution of the `tests/conftest.py/e2e_bot` fixture (which trains `e2ebot`), it takes ~37sec with `main` and ~94sec with `tf-2.5` (both repeated a few times). I'm now gonna dig deeper into this.

**Comment by RASADSA** [comment]
Hey @samsucik :wave: 
> I'm now picking this one up  Will let you know if there's space for help (esp. for testing on M1s), @mpomarole @vincelamm.
i also join the M1 test candidate line :raised_back_of_hand:

**Comment by samsucik** [comment]
Having run the regression tests [here](https://github.com/RasaHQ/rasa/pull/8899#issuecomment-880032330), it looks like switching to TF 2.5 indeed doesn't introduce regressions. BUT:
- it takes much longer to train the models (around 3x longer for the whole regression test suite to finish)
- three training runs on the Hermit dataset (all the ones using DIET with transformer) [have exited with error code 137](https://github.com/RasaHQ/rasa/actions/runs/1025512513) (I'll re-run these to see if it happens again)

As a limitation of our current regression tests, we don't track memory usage and NLU training time, and so these two should be checked some other way (as mentioned, the training time went up).

**Comment by dakshvar22** [comment]
@samsucik Do we have a small tensorflow code snippet with a very basic model and dummy data that could demonstrate this increase in training time? I think it would be worthwile to ask the tensorflow folks if we are doing something wrong or the increase is indeed expected.

Also, is the increase in time observable on both CPU and GPU?

**Comment by samsucik** [comment]
@dakshvar22 I'm coming back to this after almost a week and my previous comments basically summarise where I had left off. What you're describing is the state I'd like to reach asap 😉 

> Also, is the increase in time observable on both CPU and GPU?

Yes. I saw the regression tests taking ~3x longer (on GPU runners) and I saw the same/similar thing with `e2ebot` trained locally on a CPU.

**Comment by samsucik** [comment]
**Runtimes & failing model regression tests**

To complete my [previous update on model regression tests](https://github.com/RasaHQ/rasa/issues/7619#issuecomment-884070114), I've re-run the tests on the Hermit dataset and tests involving the transformer version of DIET (i.e. `DIET(seq`) [all failed again](https://github.com/RasaHQ/rasa/actions/runs/1055607864), the same way as before. It looks as if the failed runs were killed because they were taking too long -- they got killed after ~2 hours, whereas they normally take only around 15 minutes (see [the most recent scheduled run](https://github.com/RasaHQ/rasa/runs/3099257896?check_suite_focus=true)). This needs further investigation -- it might be due to something hanging up, though I wouldn't necessarily link it to [the hanging up we observed previously](https://github.com/tensorflow/tensorflow/issues/46511).

All in all, **with TF 2.5, our regression test runs take ~5x longer and sometimes even get killed in the process** ([TF 2.5 run](https://github.com/RasaHQ/rasa/actions/runs/1025512513) vs [scheduled run](https://github.com/RasaHQ/rasa/actions/runs/1043356068)).

**Comment by samsucik** [comment]
**Overall status update & handover notes**

My work is in the `tf-2.5` branch and [this associated PR](https://github.com/RasaHQ/rasa/pull/9123), the old `tf-2.4` branch and the associated [PR](https://github.com/RasaHQ/rasa/pull/8899) are now behind and abandoned -- though the regression tests were run on the old PR (the new one didn't exist yet).

Having observed the regression tests taking so long, I profiled the training of our small `e2ebot` and took it from there (since this training was also taking much longer with TF 2.5). Using `snakeviz`, I noticed that a number of ops were taking particularly longer, especially `IfGrad` and `WhileGrad`, and the various TF CRF methods such as `crf_log_norm` or `crf_log_likelihood`. However, this later turned to be _slightly_ misleading because the high runtime isn't really due to something like `IfGrad`, but potentially to some of its sub-calls. Still, seeing the CRF methods taking so long, I returned to the distilled example of _our_ CRF layer used [here](https://github.com/tensorflow/tensorflow/issues/46511), [adapted it](https://github.com/RasaHQ/research/blob/44b3c281649c6733998ee3deaebc90a28f220724/projects/comparing-tf-2.3-2.5/comparison_examples.py#L14) and used it for further investigation. Indeed, I saw the CRF methods still taking longer with TF 2.5, but I found it hard to actually pinpoint the precise issue (whether by searching through others' complaints on the internet, or by digging into the code, or by using the profiling tools). While I managed to make TF Profiler work with both the `e2ebot`-training example and with the CRF one, I'm quite new to the tool and felt like I wasn't getting too much value out of it in terms of pinpointing the precise TF ops. Importantly, in the Profiler stats, I didn't see the same ops topping the runtime list as I saw when analysing the general profiling results...

All of my code used for comparing TF 2.3 vs 2.5 runs of various Python examples is [in this project dir](https://github.com/RasaHQ/research/tree/comparing-tf-2.3-2.5/projects/comparing-tf-2.3-2.5). The README should help you with everything you need.

**Comment by samsucik** [comment]
Btw as for the "investigate failed memory leak tests.. these are flakey" task in the issue description, I didn't get to look into it. I did not do memory profiling, only runtime profiling.

Regarding the failing windows tests (e.g. [here](https://github.com/RasaHQ/rasa/actions/runs/1052560491)), in some cases these are seemingly due to memory errors, but sometimes all that we observe is that a worker crashes, so it's not all so clear. Additionally, there are some Ubuntu tests failing too (though I think that's a bit flakey and may not be related to TF 2.5). In any case, I think that memory profiling would help a lot here.

**Comment by koernerfelicia** [comment]
Related issue: https://github.com/RasaHQ/rasa/issues/9129. This upgrade blocks our ability to fix since `gelu` was moved to `tensorflow core` in `2.4` and up

**Comment by koernerfelicia** [comment]
Dug a little deeper into the model regression test outputs. Looks like this also affects `DIET` **without** entity recognition. See [here](https://docs.google.com/spreadsheets/d/1LrZhz_pM9y-MDJnlzqkpeK4UYFIgPF6wQ9_rIoK6E5w/edit#gid=129670259)

**Comment by koernerfelicia** [comment]
I'm not sure what we can generalise about the effect on CPU performance. After yesterday's results re: `DIET` without entity recognition I ran some CPU regression tests to see whether this is true also on CPU. See [here](https://docs.google.com/spreadsheets/d/1LrZhz_pM9y-MDJnlzqkpeK4UYFIgPF6wQ9_rIoK6E5w/edit#gid=1323015855). Will start the full suite of CPU tests to dig into this further.

**Comment by ancalita** [comment]
> On removing `@training.enable_multi_worker` ([this TODO item](https://github.com/RasaHQ/rasa/blob/044e47d671148f64634b1250d50d329c1731d4b9/rasa/utils/tensorflow/temp_keras_modules.py#L39))

@samsucik Sentry has flagged [this error](https://sentry.io/organizations/rasahq/issues/2347717493/?project=2801673&referrer=slack) recently caused by this decorator - digging a bit through [the forum](https://forum.rasa.com/t/module-tensorflow-python-keras-engine-training-has-no-attribute-enable-multi-worker/41502/13), it seems to affect users who are upgrading to TF versions higher than the one currently specified in rasa `poetry.lock`. 
Are you able to look into this as part of this issue or should I create a new issue for this decorator specifically and place it in the Research inbox?

**Comment by koernerfelicia** [comment]
@ancalita (Sam is currently out and I'm working on the issue with Daksh). We'll remove it as part of the upgrade, though this might be a while (about to write an update as my next comment). I think we can restrict users to 2.3.3 and wait until the upgrade, unless this error is urgent, in which case I think we can remove it, and possibly create an issue for someone to look into how to enable distributed training. Latter is out of scope for this project, and may be better handled by Engine.

**Comment by koernerfelicia** [comment]
Issue affecting our custom CRF layer has been communicated to tensorflow [here](https://github.com/tensorflow/tensorflow/issues/51393). As far as we know there is no workaround. Waiting to hear otherwise or progress on the issue.

**Comment by ancalita** [comment]
@koernerfelicia the error is not urgent, it can wait until the upgrade.

**Comment by koernerfelicia** [comment]
>investigate crashed workers on Windows 3.6 and 3.7 during test_e2e_with_entity_evaluation (test-other-unit-tests)
 
Crashed tests on Windows seem to be due to OOM (see [here](https://rasa-hq.slack.com/archives/C01H2UCNP2T/p1632832706102800) for Slack discussion, and [here](https://github.com/RasaHQ/rasa/issues/9553) for possible solutions to the general issue of OOM on CI).

**Comment by koernerfelicia** [comment]
Tests for which we had to increase timeouts or memory threshold. Once TF addresses the issue above, we should aim to bring these timeouts back down.

test_train_persist_load_with_different_settings_non_windows
test_train_persist_load_with_different_settings
test_train_persist_load_with_only_entity_recognition
test_train_persist_load_with_composite_entities
test_lm_featurizer_shape_values_train
test_lm_featurizer_number_of_sub_tokens_process
TestNLULeakManyEpochs

**Comment by m-vdb** [comment]
Closing as this has been done as part of https://github.com/RasaHQ/rasa/pull/9649
